File: stvs/src/test/java/edu/kit/iti/formal/stvs/logic/io/xml/TestUtils.java
package edu.kit.iti.formal.stvs.logic.io.xml;

import edu.kit.iti.formal.stvs.StvsApplication;

import java.io.File;
import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

/**
 * @author Benjamin Alt
 */
public class TestUtils {

  public enum FileType {
    SESSION, CONCRETE, CONSTRAINT, CONFIG
  }

  public enum Status {
    VALID, ALL
  }

  private static final Map<FileType, String> prefixes;
  private static final Map<Status, Pattern> statuses;

  static {
    prefixes = new HashMap<>();
    prefixes.put(FileType.SESSION, "session");
    prefixes.put(FileType.CONCRETE, "concrete_spec");
    prefixes.put(FileType.CONSTRAINT, "constraint_spec");
    prefixes.put(FileType.CONFIG, "config");
    statuses = new HashMap<>();
    statuses.put(Status.VALID, Pattern.compile("(?!.*invalid.*).*valid.*"));
    statuses.put(Status.ALL, Pattern.compile(".*"));
  }

  public static String removeWhitespace(String input) {
    return input.replaceAll("\\s+", "");
  }

  public static List<File> getTestFiles(FileType type, Status status) throws URISyntaxException {
    List<File> res = new ArrayList<>();
    for (File testSet : getTestSets()) {
      String[] children = testSet.list();
      if (children != null) {
        for (String childName : children) {
          if (childName.startsWith(prefixes.get(type)) && statuses.get(status).matcher(childName)
              .matches()) {
            res.add(new File(testSet.getAbsolutePath() + File.separator + childName));
          }
        }
      }
    }
    return res;
  }

  private static List<File> getTestSets() throws URISyntaxException {
    File testSetsDirectory = new File(StvsApplication.class
        .getResource("testSets").toURI());
    List<File> res = new ArrayList<>();
    for (String childName : testSetsDirectory.list()) {
      res.add(new File(testSetsDirectory.getAbsolutePath() + File.separator + childName));
    }
    return res;
  }

  public static String readFromFile(String path) throws IOException {
    return new String(Files.readAllBytes(Paths.get(path)), "utf-8");
  }

  public static void main(String[] args) throws URISyntaxException {
    for (FileType type : FileType.values()) {
      for (File file : getTestFiles(type, Status.ALL)) {
        System.out.println(file.getName());
      }
    }
  }
}

File: stvs/src/test/kotlin/edu/kit/iti/formal/stvs/logic/io/xml/TestUtils.kt
/* *****************************************************************
 * This file belongs to verifaps-lib (https://verifaps.github.io).
 * SPDX-License-Header: GPL-3.0-or-later
 * 
 * This program isType free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program isType distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a clone of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * *****************************************************************/
package edu.kit.iti.formal.stvs.logic.io.xml

import edu.kit.iti.formal.stvs.StvsApplication
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.OutputStream
import java.net.URISyntaxException
import java.nio.file.Paths
import kotlin.io.path.bufferedReader

/**
 * @author Benjamin Alt
 */
object TestUtils {
    private val prefixes =
        hashMapOf(
            FileType.SESSION to "session",
            FileType.CONCRETE to "concrete_spec",
            FileType.CONSTRAINT to "constraint_spec",
            FileType.CONFIG to "config",
        )

    private val statuses: HashMap<Status, Regex> = hashMapOf(
        Status.VALID to "(?!.*invalid.*).*valid.*".toRegex(),
        Status.ALL to ".*".toRegex(),
    )

    fun removeWhitespace(input: String) = input.replace("[\t\n\r ]+".toRegex(), "").replace("\r\n", "\n").trim()

    fun getResource(name: String) = javaClass.getResourceAsStream(name) ?: error("Could not find resource file: $name")

    @Throws(URISyntaxException::class)
    fun getTestFiles(type: FileType, status: Status): List<File> {
        val res: MutableList<File> = ArrayList()
        for (testSet in testSets) {
            val children = testSet.list()
            if (children != null) {
                for (childName in children) {
                    if (childName.startsWith(prefixes[type]!!) &&
                        statuses[status]!!.matches(childName)
                    ) {
                        res.add(File(testSet.absolutePath + File.separator + childName))
                    }
                }
            }
        }
        return res
    }

    @get:Throws(URISyntaxException::class)
    private val testSets: List<File>
        get() {
            val testSetsDirectory = File(
                StvsApplication::class.java.getResource("testSets")!!.toURI(),
            )
            val res: MutableList<File> = ArrayList()
            for (childName in testSetsDirectory.list()!!) {
                res.add(File(testSetsDirectory.absolutePath + File.separator + childName))
            }
            return res
        }

    fun readFromFile(path: String) = Paths.get(path).bufferedReader().readText()

    fun stringOutputStream(block: (OutputStream) -> Unit): String {
        val out = ByteArrayOutputStream(4096)
        block(out)
        return out.toString("utf-8")
    }

    enum class FileType {
        SESSION,
        CONCRETE,
        CONSTRAINT,
        CONFIG,
    }

    enum class Status {
        VALID,
        ALL,
    }
}